### ISSUE TITLE
Add server/client configuration to port option (like with "host" option)

### ISSUE THREAD

**Issue by DBosley** [webpack-contrib/webpack-hot-client#79]
* Operating System: Ubuntu 18.04
* Node Version: 8.11.1
* NPM Version: 6.0.0
* webpack Version: 4.12.0
* webpack-hot-client Version: 4.0.3

- [ ] This is a **bug**
- [x] This is a **feature** request
- [ ] This is a **modification** request

### Code

```js
// Suggested Configuration:
{
      host: {
        client: "my.website.com",
        server: "0.0.0.0"
      },
      https: true,
      port:{
        client: false, // number or false
        server: 8081
      }
    }
```

### For Features; What is the motivation and/or use-case for the feature?
This would allow users who have a proxy configuration set up to allow the default port 443 for wss:// connections to work as intended. The issue I'm having is that the address the websocket tries to listen over is `wss://my.website.com:8081` which means I have to set my nginx server running on docker to also listen to port 8081 as well as 443. The ideal way this would work for me would be that the client just tries to connect to `wss://my.website.com` without any port information added. Then it would route correctly through 443

**Comment by shellscape** [comment]
Thanks for the issue :beer:

This has been requested (once?) before, early on when `webpack-serve` was in its infancy. The case was made for allowing the hosts to be different, which were previously enforced because the added complexity and unpredictability of the client experience outweighed the benefit. When calls for that feature made it clear that the benefit outweighed the afore mentioned potential problems, we made the change. 

This is one of those situations where the use-case is super-niche. (From what I recall, as was the first time it was suggested). Given that there's a workaround for your particular environment/setup, and the slim likelihood that this would be a widely used feature, I'm hesitant to approve of adding it. 

> which means I have to set my nginx server running on docker to also listen to port 8081 as well as 443

For a development environment, this doesn't seem an unreasonable workaround, and appears to be something easily reproducible and trackable with the right docker config and container setup. It could even be something made public for exactly this use case.

There's a high bar for adding complexity to `webpack-hot-client`, so I'm open to being convinced. But at the moment I'm just not convinced that it's the responsibility of this tool to cater to that case (possibly edge-case, definitely niche).

**Comment by DBosley** [comment]
It turns out that there is no workaround. I was incorrect about that. When I set docker to listen to port 8081 it prevents the websocket server from running on it as well.

**Comment by DBosley** [comment]
I found a hacky workaround to this. Ending the client "host" config with a `?` puts the port in the querystring for the url and then it goes to the default wss:// port (443).

**Comment by shellscape** [comment]
😆 that's extremely clever man, I must say. That must be a side-effect of `url.format` [here](https://github.com/webpack-contrib/webpack-hot-client/blob/master/lib/client/socket.js#L10)

**Comment by DBosley** [comment]
I don't like the solution, but it gets the job done for now. But I do think this problem will affect anyone who wants to run nginx in a container andthey don't want to run webpack-serve in a container as well. Ours is a fairly complex microservice architecture and our philosophy is to keep our containers prod-like so a dev server really doesn't have a place in them. I'd really love to see this config change implemented so I don't need to rely on a side-effect to use a modern front-end dev stack.

**Comment by shellscape** [comment]
> to run nginx in a container and they don't want to run webpack-serve in a container as well

Running the [dev] server in the container is probably the most common way of utilizing webpack-serve (and webpack-dev-server) with containers. 🤐 

> Ours is a fairly complex microservice architecture and our philosophy is to keep our containers prod-like so a dev server really doesn't have a place in them.

If you're using as dev-environment tool to test a production-like container, I'd argue that your architecture is just a bit off. That's not to say you're bananas or anything. A dev environment tool is just not meant to be suitable for production testing. You might benefit from using `koa` or `express` directly and using `webpack-dev-middleware` directly, cutting out all the sugar webpack-serve provides.

> There's a high bar for adding complexity to webpack-hot-client, so I'm open to being convinced.

I swear I'm not being cheeky in the slightest: So far there hasn't been much substantial evidence or argument for why this should be added outside of your environment needing it. That's enough within a company or small group setting, but with the massive ecosystem that webpack caters to, a more substantial set of reasoning for adding the feature, beyond a niche use-case, is needed. You might say that with such a massive ecosystem, there's surely to be another person that runs into this - and you'd probably be right - but that's where webpack tooling has gotten into trouble in the past; trying to appease every single request that's been made, without consideration to whether the added complexity and burden of maintenance of a new feature is worth its addition for the entire user-base.

**Comment by DBosley** [comment]
After thinking about this a bit more I see where you're coming from, but I think there's still a problem with it. If you're using a proxy to handle SSL you can't put it in front of webpack-hot-client. If I set up nginx to listen for any comms on 8081 to then direct to 8081 (because I can't configure the hot-client to request one port and listen on another), that doesn't work. Nginx takes control of the port and now I can't run webpack-hot-client on it. The only way around this would be running webpack-hot-client and my whole dev middleware stack outside of my proxy and set it up to use the certs itself, which complicates things when you're handling all other SSL communications in one place but now you need to configure your middleware stack differently. 

This also makes it so you are forced to be running the dev middleware when debugging locally. The way we've set up our dev server was to let nginx direct any requests to /static to webpack-serve if webpack-serve was running, otherwise direct them to our web server which would serve out the webpack-built static files. This is a really useful feature because it lets you verify the built files locally.

This is only one example of how having a proxy service in front of your connections can be useful, and I'm sure there are plenty more. All other aspects of webpack development that I know of support using a proxy in front of your site.

Finally, to your point:
> trying to appease every single request that's been made, without consideration to whether the added complexity and burden of maintenance of a new feature is worth its addition for the entire user-base.

I understand this completely and would agree if it was a complex change. The long and short of it is that the change is very minor, doesn't need to affect anyone who already has webpack-serve and webpack-hot-client configured, and allows more flexibility in configuration without adding any extra complexity. The PR #80 I created for this only took about 15 minutes. Granted it failed the code coverage CI, but that should be a simple fix to get it in line with your standards.

**Comment by shellscape** [comment]
> I understand this completely and would agree if it was a complex change.

It's more of a snowball effect. The more "relatively simple" features we add, the more we have to support long term. That's why we set such a high bar. 

Overall I really like your explanation. I'd like some time to consider it more thoroughly, but it makes a compelling case for why this should be included. Really appreciate your patience with this.

**Comment by DBosley** [comment]
Thank you for being open minded about it. I appreciate that as well. For now my hack gets the job done, but I'd hate to have to go back to `webpack-dev-server` if for some reason it gets patched away in a later version of node.

**Comment by xorander00** [comment]
Throwing in my own $0.02 because I'm running into this issue currently as well.  Also, this comment might be useful for others that are looking to solve this issue in the near-term.  Please note that if anything comes off as snarky, I apologize, that's not my intent.
  
It's my opinion that there will be more issues like this that pop up over the coming 6-12 months or people are going to look for other options because they're finding themselves fighting with the tooling more than working on the core of their project.  
  
[See Service Workers](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API/Using_Service_Workers) as an example.  They require HTTPS and/or a secure origin to work properly.  Both Google and Mozilla are going to start tightening their browser policies to reject requests that don't comply with stuff like that.  Self-signed certificates are not an option because there are still cases where the browser ignores run-time flags to disable certificate verification.  I have a dedicated domain + wildcard SSL/TLS certificate for development that I use.  I also do most of my development over SSH with X forwarding and vim, so a given project usually has workspaces for the various components of the project spread across several systems, all of which I control from my desktop.  The service setup is that I use HAproxy at the border for SSL/TLS-termination & then routing to the appropriate internal servers for the application.  The internal servers use nGinx as the reverse proxy to map / to locally running services as necessary, which in this case is webpack-serve.
  
So basically, I need to internally have webpack-serve bind to localhost:3000 (http) & localhost:3001 (ws).  Then I need the client application to connect to https://my.domain.com for the regular resources & wss://my.domain.com for the HMR/WebSocket connection (which will be terminated & routed from HAproxy to nGinx to webpack-serve properly).  Ideally, there should be a specific path that webpack-serve configures the client to make the HTTP request to during the handshake/upgrade process to WSS.  That would solve the issue because I can just path match it in nGinx and then proxy_pass it to the proper port (3001 in my case).  
  
I respect the standard for fully evaluating feature implementations and vetting them to avoid unneeded complexity.  However, the ecosystem is large and heterogeneous.  There are a lot of different environment setups.  At the end of the day, the better option IMO (wrong or right, just an opinion), is to allow for explicit configuration of the ip:port/path server & client configs, have sensible defaults, and a message like what's there now that says "Buyer beware! This isn't supported."  The alternative is that people get frustrated of fighting the tooling, end up ripping out webpack/webpack-serve, & then having a very bad taste left in their mouths.  I wouldn't be surprised at all if there have been a relatively non-trivial amount of people that have run into this specific issue and decided it wasn't worth their time, they probably just didn't comment on it here.
  
**TL;DR**  
- Option 1: Use the @DBosley hack (slick though, gotta admit) by appending "?" to the host.
- Option 2: Move back to webpack-dev-server if you need different host:port combos.

**Comment by shellscape** [comment]
@xorander00 thanks for adding to the conversation. you can also track discussion on #81, even though that PR has been closed. it's something we now know needs doing, and the case has been very well made by everyone involved. we're now working through how best to solve it.

**Comment by xorander00** [comment]
@shellscape Ah, I missed #81.  That's good to hear.  If I have any other thoughts, where's the preferred place to post them?  The only suggestion is potentially allowing a configurable path for the client to connect to as well to perform the upgrade handshake.

**Comment by shellscape** [comment]
There's no real "proper" place per se. Was just giving you the heads up about the other thread.

I think what we're going to do is spin up a secondary `https.Server` for the WebSocket server to use, if and only if the host or port for the client are set other than the default. If the host and port for the client are default, then we'll probably pass the `https.Server` meant for koa along to hot-client.

**Comment by JSteunou** [comment]
Sorry to be not very constructive, but I ended up in a case where we need this feature as well at work. Our stack involve some docker and nginx as reverse proxy, and being able to set up the client / server with more independence would be a life safer :)
